crates/valence_schem/src/lib.rs

```diff
@@ -0,0 +1,1112 @@
+
+    #[error("invalid offset")]
+    InvalidOffset,
+
+    #[error("missing block palette")]
+    MissingBlockPalette,
+
+    #[error("invalid block palette")]
+    InvalidBlockPalette,
+
+    #[error(transparent)]
+    ParseBlockStateError(#[from] ParseBlockStateError),
+
+    #[error("missing block data")]
+    MissingBlockData,
+
+    #[error(transparent)]
+    VarIntDecodeError(#[from] VarIntDecodeError),
+
+    #[error("block {0} not in palette {1:?}")]
+    BlockNotInPalette(i32, HashMap<i32, BlockState>),
+
+    #[error("unknown block state id {0}")]
+    UnknownBlockStateId(i32),
+
+    #[error("invalid block count")]
+    InvalidBlockCount,
+
+    #[error("missing block entity pos")]
+    MissingBlockEntityPos,
+
+    #[error("invalid block entity pos {0:?}")]
+    InvalidBlockEntityPos(Vec<i32>),
+
+    #[error("missing block entity id")]
+    MissingBlockEntityId,
+
+    #[error("invalid block entity id '{0}'")]
+    InvalidBlockEntityId(String),
+
+    #[error("unknown block entity '{0}'")]
+    UnknownBlockEntity(String),
+
+    #[error("missing biome palette")]
+    MissingBiomePalette,
+
+    #[error("invalid biome palette")]
+    InvalidBiomePalette,
+
+    #[error("biome {0} not in palette {1:?}")]
+    BiomeNotInPalette(i32, HashMap<i32, Ident<String>>),
+
+    #[error("invalid biome ident '{0}'")]
+    InvalidBiomeIdent(String),
+
+    #[error("missing biome data")]
+    MissingBiomeData,
+
+    #[error("invalid biome count")]
+    InvalidBiomeCount,
+
+    #[error("missing entity pos")]
+    MissingEntityPos,
+
+    #[error("invalid entity pos {0:?}")]
+    InvalidEntityPos(Vec<f64>),
+
+    #[error("missing entity id")]
+    MissingEntityId,
+
+    #[error("invalid entity id '{0}'")]
+    InvalidEntityId(String),
+}
+
+struct VarIntReader<I: ExactSizeIterator<Item = u8>>(I);
+impl<I: ExactSizeIterator<Item = u8>> Iterator for VarIntReader<I> {
+    type Item = Result<i32, VarIntDecodeError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        struct ReadWrapper<I: ExactSizeIterator<Item = u8>>(I);
+        impl<I: ExactSizeIterator<Item = u8>> Read for ReadWrapper<I> {
+            fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+                for (idx, byte) in buf.iter_mut().enumerate() {
+                    let Some(val) = self.0.next() else {
+                        return Ok(idx);
+                    };
+                    *byte = val;
+                }
+                Ok(buf.len())
+            }
+        }
+
+        if self.0.len() == 0 {
+            None
+        } else {
+            Some(VarInt::decode_partial(ReadWrapper(&mut self.0)))
+        }
+    }
+}
+
+struct VarIntWriteWrapper<'a>(&'a mut Vec<i8>);
+impl<'a> Write for VarIntWriteWrapper<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.0.extend(buf.iter().map(|byte| *byte as i8));
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
```

Comment:
I'm going to expose some functions in `valence_nbt` that should make these wrappers unnecessary.